### ISSUE TITLE
off-centered accuracy ring

### ISSUE THREAD

**Issue by cammace** [mapbox/mapbox-gl-native#4929]
Tilting the mapview while `myLocationEnabled` will result in the accuracy ring becoming off-centered around the user dot. 

![screenshot_20160502-124843](https://cloud.githubusercontent.com/assets/5652865/14960869/cdd1f87e-1064-11e6-8d0f-c1ab4c4cdb1e.png)

cc: @tobrun

**Comment by tobrun** [comment]
Refs #4897

**Comment by tobrun** [comment]
I'm currently using the following method to draw the circle:
```
 canvas.drawCircle(foregroundBounds.centerX(), foregroundBounds.centerY(), accuracyPixels <= maxRadius ? accuracyPixels : maxRadius, accuracyPaint);
```

I need to look into which bounds I should be using instead of the foregroundBounds.

**Comment by tobrun** [comment]
It seems the solution is not as easy as I initially thought above. 
See this screenshot were I added a color to the background. 

![device-2016-05-04-134801](https://cloud.githubusercontent.com/assets/2151639/15013299/1b884208-11ff-11e6-985b-d76cc0389d9c.png)

**Comment by tobrun** [comment]
After some debugging I'm noticing that bounds are not changed when applying rotationX

**Comment by bleege** [comment]
Just testing this out with the latest code from the `release-android-v4.1.0` branch and seeing that the problem still exists.

![device-2016-05-27-141006](https://cloud.githubusercontent.com/assets/989928/15618627/c6cabaf2-2414-11e6-9222-70359417708b.png)

**Comment by tobrun** [comment]
This issue comes from the following line:

```java
canvas.drawCircle(foregroundBounds.centerX(), foregroundBounds.centerY(),....);
```

When not transformed. The above statement is working as designed. The problem originates if we apply `setRotationX` on the whole View (rotation over the x-axis). This transformation changes how the dot will be drawn but not calculates in the difference when drawing the circle. The `foregroundBounds.centerY()` resembles the non transformed y-coordinate.

**Comment by tobrun** [comment]
Instead of drawing the circle ourselves, I created a drawable for it. Now you can see the rotationX is applied when tilted. 

<img width="263" alt="screen shot 2016-05-30 at 11 17 11" src="https://cloud.githubusercontent.com/assets/2151639/15645346/26b28f36-2658-11e6-8d0e-f391353652df.png">

While a fixed size drawable does the trick, it seems we are back at square one. We need to be able to scale the drawable based on the accuracy value returned by the GPS. Scaling on a canvas depends on topleft of a View or a provided coordinate. The latter one's are the same coordinates required  for a correct `drawCircle` implementation from above.

**Comment by tobrun** [comment]
If have been trying different approaches to solve this issue. From using `Camera`, transformation matrices to calculating the offset manually. After almost a day of not making progress I'm going to punt on this and will revisit this later.

**Comment by tobrun** [comment]
This [SO](http://stackoverflow.com/a/27844491/1281750)-post talks about the ability to set the Android SDK Camera distance.This distance passed in will be used to calculate the transformation. I'm thinking that this will be the missing piece to calculate the correct x,y coordinate using that Camera API. I'm currently not sure how I would convert our current z value (zoom level) to pixels.

**Comment by tobrun** [comment]
With above in place I'm seeing the behaviour I was expecting, only the value for camera distance is not correct which results in a too large y translation. 

@1ec5 would it be possible to convert a zoom level to an camera distance?

**Comment by 1ec5** [comment]
The iOS and OS X SDKs have [some code to convert between zoom levels and altitudes](https://github.com/mapbox/mapbox-gl-native/blob/b94a125ff28db2d4bfd181928ce8eee523a30ec7/platform/darwin/src/MGLGeometry.mm#L33-L46). At the time I wrote it, I avoided putting it in mbgl because I didn’t realize the Android SDK needed it too. The calculation matches MapKit but is technically inconsistent with mbgl (but still “good enough” for most purposes).

Note that the iOS SDK doesn’t use these conversion routines for positioning or sizing the user dot or its accuracy ring. For that, we [size the ring based on the current meters per point](https://github.com/mapbox/mapbox-gl-native/blob/b94a125ff28db2d4bfd181928ce8eee523a30ec7/platform/ios/src/MGLUserLocationAnnotationView.m#L529-L530), then [rotate both elements together based on the degrees of pitch](https://github.com/mapbox/mapbox-gl-native/blob/b94a125ff28db2d4bfd181928ce8eee523a30ec7/platform/ios/src/MGLUserLocationAnnotationView.m#L163).

**Comment by tobrun** [comment]
@1ec5 thank you for linking that. I tried porting it first to java and see if it resolved the issue. At this point I'm not able to get calculate the correct value. I don't think it related to the linked code but the way I'm setting up the Android SDK Camera object and retrieving the transformation matrix from it.

> Note that the iOS SDK doesn’t use these conversion routines for positioning or sizing the user dot or its accuracy ring. 

That was my initial idea as well. I'm now thinking of rewriting the `onDraw` logic so we also don't need to explicitly calculate that position. This will be a big lift but I feel it's better to put time into that than try working around the issue. It's cleaner and will be performant because it doesn't require the extra calculations.

**Comment by tobrun** [comment]
I have rewritten the way we setup and draw the content:

![ezgif com-video-to-gif 26](https://cloud.githubusercontent.com/assets/2151639/15677821/366f51bc-274c-11e6-9e49-9c4e2884fda9.gif)

Instead of leveraging `setRotationX`, I'm using an own created Android SDK Camera object and performing the rotation over the x axis myself.

**Comment by tobrun** [comment]
While looking into merging the related PR, I noticed that current proposed implementation introduced regressions when it comes to location updates. When not receiving a location update everything is working as designed but when a location update occurs the View location doesn't match the update and breaks showing the location completely. Currently looking into resolving this but not able to give an eta when it will be resolved.

**Comment by tobrun** [comment]
I was finally able to pinpoint the issue to rotation. The rotation animation we are applying now impacts the viewport of MyLocationView and combining this with Android SDK camera object results in a wrong perspective. Next steps is to integrate rotation inside of the camera approach and fixup the remaining items in the PR https://github.com/mapbox/mapbox-gl-native/pull/5192#issuecomment-222915168

**Comment by bleege** [comment]
Rebased and Merged into `release-android-v4.1.0`.  Great job @tobrun!
